src/ringct.rs

```diff
@@ -37,23 +37,51 @@ pub struct RingCtMaterial {
 impl RingCtMaterial {
     pub fn sign(
         &self,
-        msg: &[u8],
         pc_gens: &PedersenGens,
         mut rng: impl RngCore,
-    ) -> Result<(RingCtTransaction, Vec<RevealedCommitment>)> {
+    ) -> Result<(Vec<u8>, RingCtTransaction, Vec<RevealedCommitment>)> {
         let bp_gens = BulletproofGens::new(RANGE_PROOF_BITS, RANGE_PROOF_PARTIES);
         let mut prover_ts = Transcript::new(MERLIN_TRANSCRIPT_LABEL);
 
-        // We create a ring signature for each input
-        let mut mlsags = Vec::new();
-        let mut revealed_pseudo_commitments = Vec::new();
-        for mlsag_material in self.inputs.iter() {
-            let (mlsag, revealed_pseudo_commitment) = mlsag_material.sign(msg, &pc_gens, &mut rng);
-            mlsags.push(mlsag);
-            revealed_pseudo_commitments.push(revealed_pseudo_commitment)
-        }
+        // We need to gather a bunch of things for our message to sign.
+        //   All public keys in all rings
+        //   All key-images,
+        //   All PseudoCommitments
+        //   All output commitments
+        //   All output range proofs
+
+        // All public keys in all rings
+        let true_input_public_keys: Vec<G1Projective> = self
+            .inputs
+            .iter()
+            .map(|m| m.true_input.public_key())
+            .collect();
+        let decoy_inputs_public_keys: Vec<G1Affine> = self
+            .inputs
+            .iter()
+            .flat_map(|m| {
+                m.decoy_inputs
+                    .iter()
+                    .map(|d| d.public_key())
+                    .collect::<Vec<G1Affine>>()
+            })
+            .collect();
 
-        // Now prepare the output commitments
+        // All key-images,
+        let true_input_key_images: Vec<G1Projective> = self
```

Comment:
It's enough to just say `key_images`. All key images are from the true inputs.
```suggestion
        let key_images: Vec<G1Projective> = self
```